spec/unit/webrtc/groupCall.spec.ts

```diff
@@ -14,85 +14,305 @@ See the License for the specific language governing permissions and
+
+    getUserId() { return this.userId; }
+
+    getDeviceId() { return this.deviceId; }
+    getSessionId() { return this.sessionId; }
+
+    emit = jest.fn();
+    on = jest.fn();
+    removeListener = jest.fn();
+    getTurnServers = () => [];
+    isFallbackICEServerAllowed = () => false;
+    reEmitter = new ReEmitter(new TypedEventEmitter());
+    getUseE2eForGroupCall = () => false;
+    checkTurnServers = () => null;
+}
 
 describe('Group Call', function() {
     beforeEach(function() {
-        // @ts-ignore Mock
-        global.AudioContext = MockAudioContext;
+        installWebRTCMocks();
     });
 
-    it("sends state event to room when creating", async () => {
-        const mockSendState = jest.fn();
+    describe('Basic functionality', function() {
+        let mockSendState;
+        let mockClient;
+
+        beforeEach(function() {
+            const typedMockClient = new MockCallMatrixClient(
+                FAKE_USER_ID_1, FAKE_DEVICE_ID_1, FAKE_SESSION_ID_1,
+            );
+            mockSendState = typedMockClient.sendStateEvent;
+
+            mockClient = typedMockClient as unknown as MatrixClient;
+        });
+
+        it("sends state event to room when creating", async () => {
+            const room = new Room(FAKE_ROOM_ID, mockClient, FAKE_USER_ID_1);
+            const groupCall = new GroupCall(mockClient, room, GroupCallType.Video, false, GroupCallIntent.Prompt);
+
+            await groupCall.create();
+
+            expect(mockSendState.mock.calls[0][0]).toEqual(FAKE_ROOM_ID);
+            expect(mockSendState.mock.calls[0][1]).toEqual(EventType.GroupCallPrefix);
+            expect(mockSendState.mock.calls[0][2]["m.type"]).toEqual(GroupCallType.Video);
+            expect(mockSendState.mock.calls[0][2]["m.intent"]).toEqual(GroupCallIntent.Prompt);
+        });
+
+        it("sends member state event to room on enter", async () => {
+            const room = new Room(FAKE_ROOM_ID, mockClient, FAKE_USER_ID_1);
+            const groupCall = new GroupCall(mockClient, room, GroupCallType.Video, false, GroupCallIntent.Prompt);
+
+            room.currentState.members[FAKE_USER_ID_1] = {
+                userId: FAKE_USER_ID_1,
+            } as unknown as RoomMember;
+
+            await groupCall.create();
+
+            try {
+                await groupCall.enter();
+
+                expect(mockSendState.mock.lastCall[0]).toEqual(FAKE_ROOM_ID);
+                expect(mockSendState.mock.lastCall[1]).toEqual(EventType.GroupCallMemberPrefix);
+                expect(mockSendState.mock.lastCall[2]['m.calls'].length).toEqual(1);
+                expect(mockSendState.mock.lastCall[2]['m.calls'][0]["m.call_id"]).toEqual(groupCall.groupCallId);
+                expect(mockSendState.mock.lastCall[2]['m.calls'][0]['m.devices'].length).toEqual(1);
+                expect(
+                    mockSendState.mock.lastCall[2]['m.calls'][0]['m.devices'][0].device_id,
+                ).toEqual(FAKE_DEVICE_ID_1);
```

Comment:
I wonder if one big `expect(mockSendState).toHaveBeenLastCalledWith(...)` would be more readable

Reply:
Yeah, probably - done